Add D3 Histogram for Time Conversion Histogram

alexkim205 opened the pull request:
## PR Checklist

- [X] A custom (and reusable ♻️) D3 histogram component that is used in our time conversion display.
~~- [ ] Horizontal and vertical layouts~~
~~- [ ] Visual tweaks and animation~~

### To be addressed in other PR's

- [ ] Group outliers (@Twixes is this built out in the backend right now?)
- [ ] Bin counts and autobinning (related to #4995) 
- [ ] All steps (blocked on #5110 being merged in)
- [ ] Accuracy of data partially blocked on #5116.
- [ ] Unordered steps (I think this is still blocked on backend dev)
- [ ] Conversion window
- [ ] Empty states
- [ ] Need to move d3 logic into kea

It was difficult to mimic the exact look of our histogram [Figma](https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=498%3A39) mocks using `<LineGraph/>` so I decided to implement a histogram chart with D3. This is done through a custom `useD3` hook that lets me manipulate DOM elements using plain JS.

Part of #4855.

**Before - Line Graph**

<img width="1189" alt="Screen Shot 2021-07-13 at 12 21 23 AM" src="https://user-images.githubusercontent.com/13460330/125408950-48abe800-e370-11eb-829c-dc2870203872.png">


**After - custom D3 graph**

https://user-images.githubusercontent.com/13460330/125537930-80c901d9-d79f-4689-bd0f-d8f81cf77fa0.mov


**Action Header Bar**

<img width="787" alt="Screen Shot 2021-07-13 at 9 36 47 PM" src="https://user-images.githubusercontent.com/13460330/125561957-019f598a-2554-4da8-b31a-39acd41bfc42.png">


## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

alexkim205 committed:
add d3 histogram

alexkim205 committed:
different layouts and fine tuning d3 histogram

alexkim205 committed:
spacing math and comments

clarkus reviewed (approved):
These changes look good to me. This might also partially address https://github.com/PostHog/posthog/issues/2656

alexkim205 committed:
row-align steps action dropdown and refactor

mariusandra commented on frontend/src/scenes/insights/Histogram/Histogram.tsx:
```diff
@@ -0,0 +1,168 @@
+                        .append('svg:svg')
+                        .attr('viewBox', `0 0 ${config.width} ${config.height}`)
+                        .append('svg:g')
+                        .classed(isVertical ? FunnelLayout.vertical : FunnelLayout.horizontal, true)
```
nit: couldn't this be `.classed(layout, true)` to better match the search `.classed(layout)` below?

alexkim205 replied:
very true 👍

mariusandra commented on frontend/src/scenes/insights/Histogram/Histogram.tsx:
```diff
@@ -0,0 +1,168 @@
+
+                // if class doesn't exist on svg>g, layout has changed. after we learn this, reset
+                // the layout
+                const layoutChanged = !_svg.classed(layout)
+                _svg.attr('class', null).classed(layout, true)
```
This is fine for now, but we'll need to figure out a good universal way to track changes between values in logics and d3. There might even be a way to use selectors and/or actions directly here. 🤔

alexkim205 replied:
Yup this was kind of hacky and it'd be nice to put it in some kind of `histogramLogic` or something. I just kept it here bc there wasn't much state being passed around. Maybe we can refactor this later ad-hoc

mariusandra reviewed (commented):
Definitely looks good and agreed with the direction. I'm even able to understand the time scale on these graphs, less so in the "before" group :).

However I tried running it with postgres and got:

```
"Uncaught TypeError: Cannot read property 'bin1' of undefined
```

.. and the screen went white. Seems like this is coming from inside the d3 hook. I'd suggest wrapping the component in a React error boundary to prevent any future bugs there from doing more damage than they should.

There are also obvious console.log statements and PR itself is still as draft, so let me know if/when you want a deeper review... and I'll fire up the old houseclicker.

If the feature doesn't work with Postgres, best to disable it as well.

[This line in this PR](https://github.com/PostHog/posthog/pull/5077/files#diff-92e32501a6b77e6bdff549cb4f721a309c1b5fd3df14615dc6969315b66041dcR311) provides a way to check it easily.

paolodamico commented:
Will wait for @mariusandra's comments to be addressed as I can't test locally. Also, it's marked as draft so not sure if we jumped the gun here.

From the recording you uploaded, I would suggest considering dropping the two visualization options (left-to-right and top-to-bottom) for this view. I don't see a strong case for having the top-to-bottom and feels like it can introduce cognitive burden to users, one more thing to pay attention to and decide on, quoting Marius quoting Steve K, "Don't make me think". CC @clarkus for thoughts too.

mariusandra commented:
I'm still not able to test this locally, as I don't have any funnel returning any data now. However the empty view here could also be improved somewhat. At least by aligning the axes.
![image](https://user-images.githubusercontent.com/53387/125615924-232ae2fc-114d-4a44-aac3-4d37e91d0599.png)

mariusandra commented:
Also, I tend to agree with Paolo here regarding the two choices. Also, I'd personally be in favour of removing the animations. I think they're more distracting than useful TBH. E.g. in the video the bars are animating, but the timescale changes, so it's not really providing anything visually... except just slowing the chart half a second later.

mariusandra commented:
Yeah, the animations are more distracting than useful :).

![2021-07-14 16 00 14](https://user-images.githubusercontent.com/53387/125635208-f44dad0c-fc09-4975-8c90-d4bcd23bd89a.gif)

alexkim205 committed:
cleanup

clarkus commented:
> From the recording you uploaded, I would suggest considering dropping the two visualization options (left-to-right and top-to-bottom) for this view. I don't see a strong case for having the top-to-bottom and feels like it can introduce cognitive burden to users, one more thing to pay attention to and decide on, quoting Marius quoting Steve K, "Don't make me think". CC @clarkus for thoughts too.

Agreed with presenting a single option. I think most often, the left-to-right layout of histograms is most valuable. I also think dropping the animations is a good call for now. They can sometimes make a chart feel like it's performing poorly in terms of rendering time.

alexkim205 commented:
I just wanted to collect preliminary thoughts in this draft to see if I was going in the right direction, so thank you for all the helpful feedback!

re: postgres not working

From what I can see, it doesn't look like the new time conversion api is supported by postgres so definitely makes sense to hide this feature behind the `clickhouseFeatures` value that Sam created.

re: vertical and horizontal layout

Initially I added both layouts so that the time conversion experience would be consistent w/ the funnel steps, but I see that the reason why we added the horizontal layout in the first place (to fit funnels with _many_ steps) doesn't really apply here. I'll keep the dual layout functionality in the generic `<Histogram/>` component since it might be useful in different situations (maybe can apply something similar to a custom bar chart), but remove the option to change layouts in the time conversion view.

re: animations

Got it! functionality > 👁️ 🍬

alexkim205 committed:
hide time conversion behind ch flag and remove layout toggle

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/time-conversion-analysis-view

alexkim205 committed:
remove animation and layout toggle

alexkim205 commented:
> I'm still not able to test this locally, as I don't have any funnel returning any data now. However the empty view here could also be improved somewhat. At least by aligning the axes.

I'll address this in a followup PR dedicated to funnel time conversion empty states

mariusandra commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -106,9 +107,9 @@ export const funnelLogic = kea<funnelLogicType<TimeStepOption>>({
     loaders: ({ props, values, actions }) => ({
         results: [
-            [] as FunnelStep[],
+            [],
```
Why remove this?

alexkim205 replied:
Thanks! looks like i fat fingered something here. I'll add it back

mariusandra reviewed (approved):
Looks good! I'm not sure why this one type in `funnelLogic` was removed though. Can you add it back if it was a mistake?

alexkim205 committed:
revert fat finger

samwinslow reviewed (commented):
Code LGTM but it's confusing that there's no unit listed for the x-axis. Also, nitpick but it would be nice to see a border-radius on this bar to match the step bars.
 edit: disregard as I was looking at an outdated commit.

alexkim205 committed:
erroneous type
